enhancements/operator-conditions-lib.md

```diff
@@ -49,40 +49,99 @@ Operators need to be able to read the Status CR created by OLM, and also have th
+	// exist.
+  GetCondition(context.Context, metav1.ConditionType) (*metav1.Condition, error)
+
+  // SetConditionStatus sets the specific condition on the operator's 
+	// OperatorCondition to the provided status. It returns an error if there
+	// are problems getting or updating the OperatorCondition object.
+  SetConditionStatus(context.Context, metav1.ConditionType, metav1.ConditionStatus, ...Option) error
+
+  // GetNamespacedName returns the NamespacedName of the CR. It returns an error
+  // when the name of the CR cannot be found from the environment variable set by
+  // OLM, or when the namespace cannot be found from the associated service account
+  // secret.
+  GetNamespacedName(context.Context) (types.NamespacedName, error)
+}
+```
 
-In order to get the `Status` CRD from cluster using controller-runtimes's [client][cr_client], users will need [`ObjectKey`][cr_objectkey] of type [`types.NamespacedName`][cr_namespacedname]. OLM will set the name of Status CR as an [environment variable][conditions_followup_pr] while creating the resource. Namespace of the operator can be found from the associated [service account secret][accessing_api_from_pod] available in the operators' pod. The library will utilize both of these to obtain the `NamespacedName` for the CRD.
+Optional fields like `Reason` or `Message` of the condition, can be set by passing the values to `Options`.
 
-2. Specify conditions in the CRD:
+```go
+// Option is a function that applies a change to a condition.
+// This can be used to set optional condition fields, like reasons
+// and messages.
+type Option func(*metav1.Condition)
+
+func WithReason(reason string) Option {
+	return func(c *metav1.Condition) {
+		c.Reason = reason
+	}
+}
+
+func WithMessage(message string) Option {
+	return func(c *metav1.Condition) {
+		c.Message = message
+	}
+}
+```
 
-The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom ([non-OLM supported][non-olm-supported-condition]) condition for communicating the status of the operator to users. An example Status CR is given below:
+OLM currently supports the [`Upgradable`][olm_upgradable] condition. The library will provide an interface which implements the `ConditionManager` interface and fetches the `Upgradable` condition. 
 
-```yaml
-apiVersion: operators.coreos.com/v1
-kind: Status
-metadata:
-  name: foo-operator-zdgen-asdf # RandomGen Name
-  namespace: operators
-status:
-  conditions:
-    type: Upgradeable # The name of the `foo` OLM Supported Condition.
-    status: "False"   # The operator is not ready to be upgraded.
-    reason: "migration"
-    message: "The operator is performing a migration."
-    lastTransitionTime: "2020-08-24T23:15:55Z"
+```go
+type UpgradableCondition struct {
+	conditionType metav1.ConditionType
+	manager ConditionManager
+}
+
+var _ ConditionManager = &UpgradableCondition{}
```

Comment:
`UpgradableCondition` would not implement the ConditionManager interface because its Get and Set functions would not have the `conditionType` parameter.

However, if we go with my above suggestions, this would turn into:

```suggestion
var _ Condition = &condition{}
```